### PR TITLE
Windows version check in doctor

### PR DESCRIPTION
**Pull request by eliasyishak**
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

Addresses #102607 stating that a warning should be emitted during a `flutter doctor` run to warn users on Windows versions less than 10

### CODE REVIEW

**Comment by eliasyishak** [comment]
Example return from a `systeminfo` call in a Windows environment.

```
Host Name:                 XXXXXXXXXXXX
OS Name:                   Microsoft Windows 10 Enterprise
OS Version:                10.0.19044 N/A Build 19044
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Member Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          N/A
Registered Organization:   N/A
Product ID:                XXXXXXXXXXXX
Original Install Date:     8/4/2022, 2:51:28 PM
System Boot Time:          8/10/2022, 1:03:10 PM
System Manufacturer:       Google
System Model:              Google Compute Engine
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: AMD64 Family 23 Model 49 Stepping 0 AuthenticAMD ~2250 Mhz
BIOS Version:              Google Google, 6/29/2022
Windows Directory:         C:\\Windows
System Directory:          C:\\Windows\\system32
Boot Device:               \\Device\\HarddiskVolume2
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-08:00) Pacific Time (US & Canada)
Total Physical Memory:     32,764 MB
Available Physical Memory: 17,852 MB
Virtual Memory: Max Size:  33,788 MB
Virtual Memory: Available: 18,063 MB
Virtual Memory: In Use:    15,725 MB
Page File Location(s):     C:\\pagefile.sys
Domain:                    ad.corp.google.com
Logon Server:              \\CBF-DC-8
Hotfix(s):                 7 Hotfix(s) Installed.
                           [01]: KB5013624
                           [02]: KB5003791
                           [03]: KB5012170
                           [04]: KB5016616
                           [05]: KB5014032
                           [06]: KB5014671
                           [07]: KB5015895
Hyper-V Requirements:      A hypervisor has been detected. Features required for Hyper-V will not be displayed.
```

Each line will be parsed by the windows version validator by filtering for the line that is relevant.

**Comment by eliasyishak** [comment]
cc @Jasguerrero

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import 'dart:io';
+
+import '../doctor_validator.dart';
+
+class WindowsVersionValidator extends DoctorValidator {
+  const WindowsVersionValidator() : super('Windows Version');
+  @override
+  Future<ValidationResult> validate() async {
+    final ProcessResult result = await Process.run('systeminfo', <String>[]);
+    final String resultStdout = result.stdout as String;
```
Comment (Jasguerrero) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:13]: You need to check if `ProcessResult.exitCode` was successful before going further

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        versionList.add(version);
+      }
+
+      // Reset the boolean values for the next line
+      versionText = false;
+      versionSemver = false;
+      version = null;
+    }
+
+    ValidationType windowsVersionStatus = ValidationType.missing;
```
Comment (Jasguerrero) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:71]: This can be final and set it to `ValidationType.missing` on the else

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    final List<String> versionList = <String>[];
+
+    // Use two booleans to identify when you have found
+    // the word 'version' and a version number that matches
+    // the regex pattern above; only once both are found do
+    // we report back a valid version
+    bool versionText = false;
+    bool versionSemver = false;
+    String? version;
+    for (int i = 0; i < systemInfoElements.length; i++) {
```
Comment (Jasguerrero) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:40]: nit: you can iterate directly over the `systemInfoElements` (as well as the `lineElements`) and avoid using `.elementAt`

```
for (final String curLine in systemInfoElements) {...}
```

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 1 above)
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:19]: Oh good point, I am adding a try, catch to find any general ProcessExceptions now and then also doing a check afterwords on the exit code itself

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 3 above)
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart]: nice, i meant to actually refactor this to use that notation

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+import 'package:process/process.dart';
+
+import '../base/io.dart';
+import '../doctor_validator.dart';
+
+class WindowsVersionValidator extends DoctorValidator {
+  /// Validator to be run with `flutter doctor` to check
+  /// Windows host machines if they are running supported versions,
+  /// current unsupported versions = 7, 8
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:13]: Nit, add documentation comments like these to the class declaration and not the constructor.

Follow https://dart.dev/guides/language/effective-dart/documentation#do-start-doc-comments-with-a-single-sentence-summary guidelines for structuring the comment.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      : _processManager = processManager,
+        super('Windows Version');
+
+  final ProcessManager _processManager;
+
+  @override
+  Future<ValidationResult> validate() async {
+    final ProcessResult result;
+    try {
+      result = _processManager.runSync(<String>['systeminfo']);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:24]: rather than use runSync, use run and await the result. Since many doctor validators run concurrently and runSync will block on output.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    } on ProcessException {
+      return const ValidationResult(
+        ValidationType.missing,
+        <ValidationMessage>[],
+        statusInfo:
+            'Unable to run Windows version check using built-in `systeminfo`',
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:30]: Flutter developers are unlikely to know what this means. I would loop in @christopherfujino and @Jasguerrero , but I would epect that if this check fails because we can't determine the version, we should just not say anything at all rather than worry people.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      );
+    }
+
+    if (result.exitCode != 0) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:34]: Same here

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    final List<String> versionList = <String>[];
+
+    // Use two booleans to identify when you have found
+    // the word 'version' and a version number that matches
+    // the regex pattern above; only once both are found do
+    // we report back a valid version
+    bool versionText = false;
+    bool versionSemver = false;
+    String? version;
+    for (final String curLine in systemInfoElements) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:69]: Rather than breaking down each line into individual chunks and running the regex over that, use something like String.contains on each line and pull the version info out of the match object returned.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 6 above)
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:37]: Agreed, I wasn't too sure how to handle errors/failures for the validator... i just added them to handle all errors during testing

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 8 above)
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:68]: That's a good point, I'm wondering if it would be worth ditching looping over each line and use the `multiLine` boolean on `RegExp` for matches. What are your thoughts on that approach?

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:68]: This link should show the pattern working in a demo.

https://regex101.com/r/EA9AAq/1

Using the above demo, the middle group would be our version we're interested in

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:68]: I would not use multiLine unless you potentially want a match that spans multiple lines

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:68]: If you have something that matches your line, you can just run that over the whole string and it should work

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 6 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:35]: I think for both this and exitCode != 0, we have two options:

1) show what happened to the user and ask them to file a bug because something we weren't expecting happened; or
2) do what Jonah suggested, and not say anything and not worry people (anytime there's a non-green doctor validator it causes a subset of our users a lot of anxiety)

If you want to do 1, you can just not catch the ProcessException and then explicitly throw an exception for the exitCode != 0 case; uncaught exceptions from doctor validators will be caught higher up and surfaced to the user.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 7 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:39]: Do you know in what cases `systeminfo` will return non-zero?

File: packages/flutter_tools/lib/src/doctor.dart [hunk 9]
```diff
@@ -129,6 +130,9 @@ class _DefaultDoctorValidatorsProvider implements DoctorValidatorsProvider {
         flutterRoot: () => Cache.flutterRoot!,
         operatingSystemUtils: globals.os,
       ),
+      if (platform.isWindows) WindowsVersionValidator(
```
[the remark this answers is not in the document]
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/doctor.dart:133]: I'm not aware of any versioning constraints on the other platforms, i would be more than happy to chase down an answer if you can point me to someone i can chat – after that i can look into making a more general OS validator

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 8 above)
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:68]: sounds good, editing the pattern matcher to account for that

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 7 above)
Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:39]: I can't find any examples of why this would return non-zero but just wanted to cover all basis because the following code would rely on a successful exit code

File: packages/flutter_tools/lib/src/doctor.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/doctor.dart:133]: actually, it looks like we don't have a minimum macOS version, so I guess we're good for now. Your macOS version will be transitively constrained by our supported xcode version, but we already have a validator for that.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        statusInfo: 'Exit status from running `systeminfo` was unsuccessful',
+      );
+    }
+
+    final String resultStdout = result.stdout as String;
+
+    // Regular expression pattern for identifying
+    // semantic versioned strings
+    // (ie. 10.5.4123)
+    final RegExp regex =
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:52]: I would pull this into a @visibleForTesting method that takes the stdout string and then returns `Iterable<RegExpMatch> matches`, and then you can unit test different strings

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 11]
```diff
@@ -50,7 +50,7 @@ class WindowsVersionValidator extends DoctorValidator {
     // semantic versioned strings
     // (ie. 10.5.4123)
     final RegExp regex =
-        RegExp(r'\n(OS Version:\s*)([0-9]+\.[0-9]+\.[0-9]+)(.*)\n');
+        RegExp(r'\n?(OS Version:\s*)([0-9]+\.[0-9]+\.[0-9]+)(.*)\n?');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:53]: Does adding the question mark not make this also match `'BIOS Version: `?

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:53]: I don't believe so, the question mark should be only looking for 0 or 1 newline characters before `OS`. Unless I had the wrong assumption there?

I added the question mark incase there was no new line character before the `OS Version` line on print out

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:53]: So in "BIOS Version:" won't there be 0 newline characters before "OS"?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:53]: whether i'm right or wrong, this should be test cases in the unit tests :)

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart]: I see what you mean now, I'm going to try a refactor with `multiLine` enabled so that I can specify the start of the line as a reference point

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart]: I'll build some test cases with multiline enabled when i get back. thanks for the review! 😁

**Comment by zanderso** [comment]
Oops. I fatfingered the ready for review button on mobile, but there's no way in the mobile UI to change back to draft...

**Comment by jonahwilliams** [comment]
Switch to desktop view 😁

File: packages/flutter_tools/lib/src/doctor.dart [hunk 12]
```diff
@@ -129,6 +130,9 @@ class _DefaultDoctorValidatorsProvider implements DoctorValidatorsProvider {
...
+        processManager: globals.processManager,
+      ),
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/doctor.dart:135]:
```suggestion
      if (platform.isWindows)
        WindowsVersionValidator(
          processManager: globals.processManager,
        ),
```

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 13]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:process/process.dart';
+
+import '../base/io.dart';
+import '../doctor_validator.dart';
+
+const List<String> unsupportedVersions = <String>[
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:10]:
```suggestion
/// Flutter only supports development on Windows host machines version 10 and greater.
const List<String> kUnsupportedVersions = <String>[
```

Or something like that

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  '6',
+  '7',
+  '8',
+];
+
+/// Validator to be run with `flutter doctor` to check
+/// Windows host machines if they are running supported versions.
+class WindowsVersionValidator extends DoctorValidator {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:18]:
```suggestion
/// Validator for supported Windows host machine operating system version.
class WindowsVersionValidator extends DoctorValidator {
```

Since all validators are used by doctor

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+  final ProcessManager _processManager;
+
+  /// Provide a literal string as the Regex pattern
+  /// and a string to validate and get a boolean determining
+  /// if the string has at least one match
+  static Iterable<RegExpMatch> validateString(String pattern, String str,
+      {bool multiLine = true}) {
+    final RegExp regex = RegExp(
+      pattern,
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: I don't really think this helper method is carrying its weight. First, I would move the regex string to a top level and document there what it is intended to match. Then you can replace it with `regex.allMaches(contents)`

File: packages/flutter_tools/test/general.shard/windows_version_validator_test.dart [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+/// The expected validation result object for
+/// a passing windows version test
+const ValidationResult invalidWindowsValidationResult = ValidationResult(
+  ValidationType.missing,
+  <ValidationMessage>[],
+  statusInfo: 'Unable to confirm if installed Windows version is 10 or greater',
+);
+
+void main() {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/windows_version_validator_test.dart:123]: I would add a test for non-zero exit code.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    // Regular expression pattern for identifying
+    // semantic versioned strings
+    // (ie. 10.5.4123)
+    final Iterable<RegExpMatch> matches = validateString(
+        r'^(OS Version:\s*)([0-9]+\.[0-9]+\.[0-9]+)(.*)$', resultStdout);
+
+    // Use the string split method to extract the major version
+    // and check against the [unsupportedVersions] list
+    final ValidationType windowsVersionStatus;
+    String statusInfo;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:61]:
```suggestion
    final String statusInfo;
```

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart (hunk 15 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: Oh, I think from reading @christopherfujino's comment, the idea was to embed the version matching pattern in this method to make it easy to test. 

From reading the test cases, I think it would work OK without this as a public method. But if you're not going to specifically test it you should drop this.

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: Yeah, Chris mentioned that it would be helpful to make this regex helper a method that we can unit test independently, so that we can feed it test strings as we start to encounter newer versions of the `stdout` from `systeminfo`.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: You're basically not testing this method, though, because this method itself is trivial. I would just test the regex itself

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: I was thinking that this helper method would also include lines 58-61 and return a boolean. Also the rather than passing the pattern (which is actually what you want to test), that the function would receive `resultStdout`.

Then your test can look (something) like:

```
test('validateVersionString() correctly validates the output of `systeminfo`', () {
  bool result = validateVersionString('123');
  expect(result, isFalse);
  result = validateVersionString('OS Version:   8.0.0');
  // Too old
  expect(result, isFalse);
  result = validateVersionString('OS Version:     11.1.2');
  expect(result, isTrue);
});
```

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: I think this approach is fine too, its the middle ground that isn't very useful

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:31]: I do agree with @jonahwilliams that the static method was very trivial but @christopherfujino what your test looks like is essentially what the entire validator was built to check.

I shied away from creating a regex method that returned a boolean because I didn't want to assume that there would only be one match within the `stdout`. The conditionals beginning on line 60 were added to ensure that the matches only returned 1 result AND that the major version was valid.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart]: ok, makes sense, i'm fine with you removing this function

**Review by jonahwilliams — APPROVED** [review]
LGTM
